mllib/src/main/scala/org/apache/spark/ml/feature/QuantileDiscretizer.scala

```diff
@@ -114,10 +115,10 @@ final class QuantileDiscretizer @Since("1.6.0") (@Since("1.6.0") override val ui
     splits(0) = Double.NegativeInfinity
     splits(splits.length - 1) = Double.PositiveInfinity
 
-    val distinctSplits = splits.distinct
+    val distinctSplits = splits.filter(!_.isNaN).distinct
```

Comment:
@srowen  then maybe we should, as we discussed earlier on JIRA, align with R, by having a NaN checker in approxQuantile, that is, having a NaN filter inside of approxQuantile, rather than ahead of calling approxQuantile.

We can also have a same flag for user to choose to either remove NaN values or throw an error when there is NaN in data,  although, this API change will introduce collateral impact on several existing function calls.